cgen: using incomplete `?bool` type (and possibly index overflow)

### Describe the bug

title

### Reproduction Steps

```v
struct Undefined {}

type Foo[T] = ?T | Undefined

@[params]
struct Bar {
	x Foo[bool] = Undefined{}
}

fn f(b Bar) {
	dump(b)
}

fn main() {
	f()
	f(x: true)
	f(x: Foo[bool](?bool(none)))
}
```

### Expected Behavior

successful compile

### Current Behavior

```c
==================
C:/Users/mclr/AppData/Local/Temp/v_0/test.01HPJCTQEKZCQ7Z3DHNTY0YGSP.tmp.c:497: warning: WINVER redefined
C:/Users/mclr/AppData/Local/Temp/v_0/test.01HPJCTQEKZCQ7Z3DHNTY0YGSP.tmp.c:978: error: field '_bool' has incomplete type
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.4 ec21663

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.4 57a33c3.ec21663
OS: windows, Microsoft Windows 11 Pro v22000 64-bit
Processor: 12 cpus, 64bit, little endian, 

getwd: D:\Games\Proekti\JS-TS\disjs
vexe: D:\Games\Proekti\V\v\v.exe
vexe mtime: 2024-02-13 23:26:25

vroot: OK, value: D:\Games\Proekti\V\v
VMODULES: OK, value: C:\Users\mclr\.vmodules
VTMP: OK, value: C:\Users\mclr\AppData\Local\Temp\v_0

Git version: git version 2.37.0.windows.1
Git vroot status: weekly.2024.06-68-gec21663f
.git/config present: true

CC version:
thirdparty/tcc status: thirdparty-windows-amd64 a39eb79b
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/auto_eq_methods.v
+++ b/vlib/v/gen/c/auto_eq_methods.v
@@ -77,7 +77,9 @@ fn (mut g Gen) gen_sumtype_equality_fn(left_type ast.Type) string {
 		left_arg := g.read_field(left_type, name, 'a')
 		right_arg := g.read_field(left_type, name, 'b')
 
-		if variant.sym.kind == .string {
+		if variant.typ.has_flag(.option) {
+			fn_builder.writeln('\t\treturn ((*${left_arg}).state == 2 && (*${right_arg}).state == 2) || !memcmp(&(*${left_arg}).data, &(*${right_arg}).data, sizeof(${g.base_type(variant.typ)}));')
+		} else if variant.sym.kind == .string {
 			fn_builder.writeln('\t\treturn string__eq(*${left_arg}, *${right_arg});')
 		} else if variant.sym.kind == .sum_type && !typ.is_ptr() {
 			eq_fn := g.gen_sumtype_equality_fn(typ)

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -6425,13 +6425,6 @@ fn (mut g Gen) write_types(symbols []&ast.TypeSymbol) {
 				struct_names[name] = true
 				g.typedefs.writeln('typedef struct ${name} ${name};')
 				g.type_definitions.writeln('')
-
-				opt_types := sym.info.variants.filter(it.has_flag(.option))
-				for opt_type in opt_types {
-					// Option sum types are declared as _option_xx_ptr must be declared forward
-					g.options_forward << g.typ(opt_type.clear_option_and_result().set_nr_muls(1))
-				}
-
 				g.type_definitions.writeln('// Union sum type ${name} = ')
 				for variant in sym.info.variants {
 					g.type_definitions.writeln('//          | ${variant:4d} = ${g.typ(variant.idx()):-20s}')
@@ -6440,13 +6433,14 @@ fn (mut g Gen) write_types(symbols []&ast.TypeSymbol) {
 				g.type_definitions.writeln('\tunion {')
 				for variant in sym.info.variants {
 					variant_sym := g.table.sym(variant)
-					mut var := variant.ref()
+					mut var := if variant.has_flag(.option) { variant } else { variant.ref() }
 					if variant_sym.info is ast.FnType {
 						if variant_sym.info.is_anon {
 							var = variant
 						}
 					}
-					g.type_definitions.writeln('\t\t${g.typ(var)} _${variant_sym.cname};')
+					var_type := if variant.has_flag(.option) { '${g.typ(var)}*' } else { g.typ(var) }
+					g.type_definitions.writeln('\t\t${var_type} _${variant_sym.cname};')
 				}
 				g.type_definitions.writeln('\t};')
 				g.type_definitions.writeln('\tint _typ;')

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -6425,6 +6425,13 @@ fn (mut g Gen) write_types(symbols []&ast.TypeSymbol) {
 				struct_names[name] = true
 				g.typedefs.writeln('typedef struct ${name} ${name};')
 				g.type_definitions.writeln('')
+
+				opt_types := sym.info.variants.filter(it.has_flag(.option))
+				for opt_type in opt_types {
+					// Option sum types are declared as _option_xx_ptr must be declared forward
+					g.options_forward << g.typ(opt_type.clear_option_and_result().set_nr_muls(1))
+				}
+
 				g.type_definitions.writeln('// Union sum type ${name} = ')
 				for variant in sym.info.variants {
 					g.type_definitions.writeln('//          | ${variant:4d} = ${g.typ(variant.idx()):-20s}')

--- a/vlib/v/parser/if_match.v
+++ b/vlib/v/parser/if_match.v
@@ -223,6 +223,18 @@ fn (mut p Parser) is_only_array_type() bool {
 	return false
 }
 
+fn (mut p Parser) is_match_sumtype_type() bool {
+	is_option := p.tok.kind == .question
+	name_tok := if is_option { p.peek_tok } else { p.tok }
+	next_tok_kind := if is_option { p.peek_token(2).kind } else { p.peek_tok.kind }
+	next_next_tok := if is_option { p.peek_token(3) } else { p.peek_token(2) }
+
+	return name_tok.kind == .name && !(name_tok.lit == 'C' && next_tok_kind == .dot)
+		&& (((ast.builtin_type_names_matcher.matches(name_tok.lit) || name_tok.lit[0].is_capital())
+		&& next_tok_kind != .lpar) || (next_tok_kind == .dot && next_next_tok.lit.len > 0
+		&& next_next_tok.lit[0].is_capital()))
+}
+
 fn (mut p Parser) match_expr() ast.MatchExpr {
 	match_first_pos := p.tok.pos()
 	mut else_count := 0
@@ -248,10 +260,7 @@ fn (mut p Parser) match_expr() ast.MatchExpr {
 			is_else = true
 			else_count += 1
 			p.next()
-		} else if (p.tok.kind == .name && !(p.tok.lit == 'C' && p.peek_tok.kind == .dot)
-			&& (((ast.builtin_type_names_matcher.matches(p.tok.lit) || p.tok.lit[0].is_capital())
-			&& p.peek_tok.kind != .lpar) || (p.peek_tok.kind == .dot && p.peek_token(2).lit.len > 0
-			&& p.peek_token(2).lit[0].is_capital()))) || p.is_only_array_type()
+		} else if p.is_match_sumtype_type() || p.is_only_array_type()
 			|| p.tok.kind == .key_fn
 			|| (p.tok.kind == .lsbr && p.peek_token(2).kind == .amp) {
 			mut types := []ast.Type{}

--- a/vlib/v/tests/option_generic_sumtype_test.v
+++ b/vlib/v/tests/option_generic_sumtype_test.v
@@ -1,0 +1,17 @@
+struct Undefined {}
+
+type Foo[T] = ?T | Undefined
+
+@[params]
+struct Bar {
+	x Foo[bool] = Undefined{}
+}
+
+fn f(b Bar) Bar {
+	return dump(b)
+}
+
+fn test_main() {
+	a := f()
+	assert a == Bar{}
+}

--- a/vlib/v/tests/option_sum_type_test.v
+++ b/vlib/v/tests/option_sum_type_test.v
@@ -11,7 +11,7 @@ fn is_opt(f Foo) bool {
 	}
 }
 
-fn main() {
+fn test_main() {
 	dump(is_opt(Foo(?Bar(none))))
 	dump(is_opt(Foo(Baz{})))
 }

--- a/vlib/v/tests/option_sum_type_test.v
+++ b/vlib/v/tests/option_sum_type_test.v
@@ -1,0 +1,23 @@
+fn is_opt(f Foo) bool {
+	return match f {
+		Baz {
+			false
+		}
+		// vfmt off
+		?Bar {
+			true
+		}
+		// vfmt on
+	}
+}
+
+fn main() {
+	dump(is_opt(Foo(?Bar(none))))
+	dump(is_opt(Foo(Baz{})))
+}
+
+struct Baz {}
+
+struct Bar {}
+
+type Foo = ?Bar | Baz

